network/openssl_privatekey.py

```diff
@@ -0,0 +1,246 @@
+    type:
+        required: false
+        default: "RSA"
+        choices: [ RSA, DSA ]
+        description:
+            - The algorithm used to generate the SSL private key
+    force:
+        required: false
+        default: False
+        choices: [ True, False ]
+        description:
+            - Should the key be regenerated even it it already exists
+    path:
+        required: true
+        description:
+            - Name of the folder in which the generated SSL private key will be written
+'''
+
+EXAMPLES = '''
+# Generate an OpenSSL private key with the default values (4096 bits, RSA)
+- openssl_privatekey: path=/etc/ssl/private/ansible.com.pem
+
+# Generate an OpenSSL private key with a different size (2048 bits)
+- openssl_privatekey: path=/etc/ssl/private/ansible.com.pem size=2048
+
+# Force regenerate an OpenSSL private key if it already exists
+- openssl_privatekey: path=/etc/ssl/private/ansible.com.pem force=True
+
+# Generate an OpenSSL private key with a different algorithm (DSA)
+- openssl_privatekey: path=/etc/ssl/private/ansible.com.pem type=DSA
+'''
+
+RETURN = '''
+size:
+    description: Size (in bits) of the SSL private key
+    returned:
+        - changed
+        - success
+    type: integer
+    sample: 4096
+type:
+    description: Algorithm used to generate the SSL private key
+    returned:
+        - changed
+        - success
+    type: string
+    sample: RSA
+filename:
+    description: Path to the generated SSL private key file
+    returned:
+        - changed
+        - success
+    type: string
+    sample: /etc/ssl/private/ansible.com.pem
+'''
+
+class PrivateKeyError(Exception):
+
+    def __init__(self, e):
+        self.errno = e.errno
+
+
+class PrivateKey(object):
+
+    def __init__(self, module):
+        self.size = module.params['size']
+        self.state = module.params['state']
+        self.name = os.path.basename(module.params['path'])
+        self.type = module.params['type']
+        self.force = module.params['force']
+        self.path = module.params['path']
+        self.changed = True
+        self.check_mode = module.check_mode
+
+    def generate(self):
+        """Generate a keypair."""
+
+        if not self.force and os.path.exists(self.path):
+            self.changed = False
+            return
+
+        self.privatekey = crypto.PKey()
+
+        if self.type == 'RSA':
+            crypto_type = crypto.TYPE_RSA
+        else:
+            crypto_type = crypto.TYPE_DSA
+
+        try:
+            self.privatekey.generate_key(crypto_type, self.size)
+            privatekey_file = open(self.path, 'w')
+            privatekey_file.write(crypto.dump_privatekey(crypto.FILETYPE_PEM, self.privatekey))
+            privatekey_file.close()
+        except TypeError:
+            e = get_exception()
+            raise PrivateKeyError(e)
+        except ValueError:
+            e = get_exception()
+            raise PrivateKeyError(e)
```

Comment:
You're not removing the private key file here, which suggests to me that TypeError/ValueError can be raised by `generate_key` but not by `dump_privatekey`.  I suggest making it explicit by splitting the code into two try blocks:

```python
try:
    self.privatekey.generate_key(crypto_type, self.size)
except (TypeError, ValueError):
    raise PrivateKeyError(get_exception())
try:
    f  = open(self.path, 'w')
    f.write(crypto.dump_privatekey(crypto.FILETYPE_PEM, self.privatekey))
    f.close()
except IOError:
    self.remove()
    raise PrivateKeyError(get_exception())
```